src/foam/dao/QueryCachingDAODecorator.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+foam.CLASS({
+  package: 'foam.dao',
+  name: 'QueryCachingDAODecorator',
+  extends: 'foam.dao.ProxyDAO',
+
+  documentation: 'Javascript DAO Decorator which adds select caching to a delegate DAO.',
+
+  constants: [
+    {
+      name: 'PURGE',
+      value: 'PURGE'
+    }
+  ],
+
+  properties: [
+    {
+      // The cache for local storage and fast access
+      name: 'cache',
+      factory: function() { return {}; }
+    }
+  ],
+
+  methods: [
+    // Put invalidates cache and is forwarded to the source.
+    function put_(x, o) {
+      this.cache = {};
+      return this.delegate.put_(x, o);
+    },
+
+    function select_(x, sink, skip, limit, order, predicate) {
+      if (
+        // Only cache selects that have limit provided
+        limit === undefined
+        // Only handle query caching for ArraySink and Projection
+        || ! (foam.dao.ArraySink.isInstance(sink) || foam.mlang.sink.Projection.isInstance(sink))
+      ) {
+        return this.SUPER(x, sink, skip, limit, order, predicate);
+      }
+
+      let self = this;
+      let key  = [order, predicate].toString();
```

Comment:
I think the key should also contain the class of the sink so that ArraySinks and Projections don't interfere with each other.

Reply:
I will add sink back to the key.
The reason i removed it was that we always load the cache from the array property of the sink returned by delegate select so it seemed seamless to share the cached elements across both sink types